plutus-benchmark/linear-vesting/src/LinearVesting.hs

```diff
@@ -0,0 +1,243 @@
+  }
+  deriving stock (Haskell.Show)
+
+$(makeLift ''VestingDatum)
+$(makeIsDataIndexed ''VestingDatum [('VestingDatum, 0)])
+
+data VestingRedeemer = PartialUnlock | FullUnlock
+
+$(PlutusTx.makeLift ''VestingRedeemer)
+$( PlutusTx.makeIsDataIndexed
+     ''VestingRedeemer
+     [('PartialUnlock, 0), ('FullUnlock, 1)]
+ )
+
+countInputsAtScript :: ScriptHash -> List TxInInfo -> Integer
+countInputsAtScript scriptHash = go 0
+ where
+  go :: Integer -> List TxInInfo -> Integer
+  go n = List.caseList' n \txIn txIns ->
+    case addressCredential (txOutAddress (txInInfoResolved txIn)) of
+      ScriptCredential vh | vh == scriptHash -> go (n + 1) txIns
+      _                                      -> go n txIns
+
+validateVestingPartialUnlock :: ScriptContext -> Bool
+validateVestingPartialUnlock ctx =
+  let
+    txInfo :: TxInfo = scriptContextTxInfo ctx
+    SpendingScript ownRef (Just (Datum datum)) = scriptContextScriptInfo ctx
+    vestingDatum :: VestingDatum = unsafeFromBuiltinData datum
+    inputs = txInfoInputs txInfo
+
+    Just ownVestingInput = List.find ((== ownRef) . txInInfoOutRef) inputs
+    resolvedOut = txInInfoResolved ownVestingInput
+    inputAddress = txOutAddress resolvedOut
+
+    ScriptCredential scriptHash = addressCredential inputAddress
+    Just ownVestingOutput =
+      List.find ((== inputAddress) . txOutAddress) (txInfoOutputs txInfo)
+    outputDatum = txOutDatum ownVestingOutput
+
+    divCeil :: Integer -> Integer -> Integer
+    divCeil x y = 1 + (x - 1) `divide` y
+
+    asset :: AssetClass =
+      vestingAsset vestingDatum
+    oldRemainingQty :: Integer =
+      assetClassValueOf (txOutValue resolvedOut) asset
+    newRemainingQty :: Integer =
+      assetClassValueOf (txOutValue ownVestingOutput) asset
+    vestingPeriodLength :: Integer =
+      vestingPeriodEnd vestingDatum - vestingPeriodStart vestingDatum
+    currentTimeApproximation :: Integer =
+      getPOSIXTime (getLowerInclusiveTimeRange (txInfoValidRange txInfo))
+    vestingTimeRemaining :: Integer =
+      vestingPeriodEnd vestingDatum - currentTimeApproximation
+    timeBetweenTwoInstallments :: Integer =
+      vestingPeriodLength `divCeil` totalInstallments vestingDatum
+    futureInstallments :: Integer =
+      vestingTimeRemaining `divCeil` timeBetweenTwoInstallments
+    expectedRemainingQty :: Integer =
+      (futureInstallments * totalVestingQty vestingDatum)
+        `divCeil` totalInstallments vestingDatum
+    PubKeyCredential beneficiaryHash =
+      addressCredential (beneficiary vestingDatum)
+   in
+    if
+      | not (txSignedBy txInfo beneficiaryHash) ->
+          traceError "Missing beneficiary signature"
+      | firstUnlockPossibleAfter vestingDatum >= currentTimeApproximation ->
+          traceError "Unlock not permitted until firstUnlockPossibleAfter time"
+      | newRemainingQty <= 0 ->
+          traceError "Zero remaining assets not allowed"
+      | newRemainingQty > oldRemainingQty ->
```

Comment:
`newRemainingQty >= oldRemainingQty`?

https://github.com/Anastasia-Labs/linear-vesting/blob/383b388c77bdf6d80242fdbe2a6b0167b42f4b2c/src/LinearVesting.hs#L166

It doesn't really matter though